prov/shm/src/smr_progress.c

```diff
@@ -200,6 +209,78 @@ static int smr_progress_iov(struct smr_cmd *cmd, struct iovec *iov,
 	return -ret;
 }
 
+static int smr_mmap_iov_copy_out(struct smr_ep *ep, struct smr_cmd *cmd,
+				 struct iovec *iov, size_t iov_count,
+				 size_t *total_len)
+{
+	char shm_name[NAME_MAX];
+	char peer_name[NAME_MAX];
+	void *mapped_ptr;
+	int peer_id, fd, ret, num;
+
+	peer_id = (int) cmd->msg.hdr.addr;
+
+	sprintf(peer_name, "%s", ep->region->map->peers[peer_id].peer.name);
+	num = sprintf(shm_name, "%s_%s_%ld", peer_name, ep->name,
+		      cmd->msg.hdr.msg_id);
+	if (num < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "generating shm file name failed\n");
+		return errno;
+	}
+	assert(num <= NAME_MAX);
+
+	fd = shm_open(shm_name, O_RDWR, S_IRUSR | S_IWUSR);
+	if (fd < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "shm_open error\n");
+		return errno;
+	}
+
+	mapped_ptr = mmap(NULL, cmd->msg.hdr.size, PROT_READ | PROT_WRITE,
+			  MAP_SHARED, fd, 0);
+	if (mapped_ptr == MAP_FAILED) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "mmap error %s\n", strerror(errno));
+		return errno;
+	}
+	ret = ofi_copy_to_iov(iov, iov_count, 0, mapped_ptr, cmd->msg.hdr.size);
+	if (ret != cmd->msg.hdr.size) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL,
+			"mmap iov copy out error\n");
+		return FI_EIO;
+	}
+	*total_len = ret;
+	munmap(mapped_ptr, cmd->msg.hdr.size);
+	close(fd);
+	shm_unlink(shm_name);
+
+	return 0;
+}
+
+static int smr_progress_mmap(struct smr_cmd *cmd, struct iovec *iov,
+			    size_t iov_count, size_t *total_len,
+			    struct smr_ep *ep, int err)
+{
+	struct smr_region *peer_smr;
+	struct smr_resp *resp;
+	int peer_id, ret;
+
+	peer_id = (int) cmd->msg.hdr.addr;
+	peer_smr = smr_peer_region(ep->region, peer_id);
+	resp = (struct smr_resp *) ((char **) peer_smr +
```

Comment:
why char **?  Is there a macro for this?

Why are we even getting peer_smr and resp?  These are local variables that we set and then lose on function exit?  Something is off here.

Reply:
I went through my PR to use char * in such case, and added macros for address translation.

This is the at receiver/target side, and it only receives `cmd`. 
resp is still at sender/initiator side. It's not sent to the receiver/target side. Therefore we need to find the resp struct on the sender/initiator side.